File: NeetCode 150/Arrays/74. Search a 2D Matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        n = len(matrix)
        m = len(matrix[0])
        l = 0
        r = ( n * m ) - 1

        while l <= r:
            mid = (l+r) // 2
            #converting mid to row and col. Always using column, Row // and col %
            row = mid // m
            col = mid % m
            
            if(target == matrix[row][col]):
                return True
            elif(target < matrix[row][col]):
                r = mid - 1
            else:
                l = mid + 1
        return False

File: NeedCode 150 (JAVA)/Binary Search/74. Search a 2D Matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int l = 0, n = matrix.length, m = matrix[0].length;
        int r = m * n -1;
        while(l<=r){
            int mid = (l+r)/2;
            //use column length to calculate row and col
            int row = mid/m, col = mid%m;
            if (target == matrix[row][col])
                return true;
            else if (target > matrix[row][col])
                l = mid + 1;
            else
                r = mid - 1;
        }
        return false;
    }
}
